File: server/serverpackets/S_CharPacks.py
# -*- coding: utf-8 -*-

from server.codes import Opcodes
from ServerBasePacket import ServerBasePacket

class S_CharPacks(ServerBasePacket):
    def __init__(self, name, clanName, type, sex, lawful,
                 hp, mp, ac, lv, str, dex, con, wis, cha,
                 intel, accessLevel, birthday):
        ServerBasePacket.__init__(self)
        self.writeC(Opcodes.S_OPCODE_CHARLIST)
        self.writeS(name)
        self.writeS(clanName)
        self.writeC(type)
        self.writeC(sex)
        self.writeH(lawful)
        self.writeH(hp)
        self.writeH(mp)
        self.writeC(ac)
        self.writeC(lv)
        self.writeC(str)
        self.writeC(dex)
        self.writeC(con)
        self.writeC(wis)
        self.writeC(cha)
        self.writeC(intel)
        self.writeC(0) # is Administrator
        self.writeD(birthday)

    def getContent(self):
        return self.getBytes()

File: ServerPackets/S_CharPacks.cc
//
// Created by 范炜东 on 2019/5/20.
//

#include "Config.h"
#include "Opcodes.h"
#include "Templates/Character.h"
#include "S_CharPacks.h"

S_CharPacks::S_CharPacks(std::shared_ptr<db::def::Character> role) : ServerPacket() {
  writeC(Opcodes::S_OPCODE_CHARLIST);
  writeS(role->character_name.c_str());
  writeS(role->clan_name.c_str());
  writeC(role->career);
  writeC(int(role->sex));
  writeH(role->justice());
  writeH(role->cur_hp());
  writeH(role->cur_mp());
  if (Config::CHARACTER_CONFIG_IN_SERVER_SIDE) {
    writeC(role->level);
  } else {
    writeC(1);
  }
  writeC(role->ac());
  writeC(role->str());
  writeC(role->dex());
  writeC(role->con());
  writeC(role->men());
  writeC(role->cha());
  writeC(role->wis());
  writeC(role->access_level());
  writeD(std::mktime(&role->birthday));
}